Feature/auto analytics adder

lennarkivimae opened the pull request.

Commit:
feat:auto analytics adder workflow

Adding interval to add analytics panel for all of the dashboards. Eases
maintenance in team environment, where dev-s don't need to think about
adding an panel manually. Interval runs once in 24 hours

Commit:
auto-analytics-adder - separate general api methods from worker

tiithansen commented on server/main.go:
```diff
@@ -73,6 +76,28 @@ func main() {
+	}
+
+	// Run once per 24 hours
+	ticker := time.NewTicker(24 * time.Hour)
```
Could make this timeout configurable, also http endpoint to trigger run would be nice addition, it would make it easier to test.

tiithansen commented on server/main.go:
```diff
@@ -73,6 +76,28 @@ func main() {
-	err := http.ListenAndServe(cli.HTTPAddress, mux)
-	ctx.FatalIfErrorf(err)
+	go func() {
+		err := http.ListenAndServe(cli.HTTPAddress, mux)
```
Isn't it enough to just move it to the end of function? Then you should not need to run it in separate goroutine.

lennarkivimae replied:
AFAIK ListenAndServe runs indefinitely. Meaning the anything after it doesn't get executed. However it should block main function from exiting. Rearranging the ticker loop and removing separate goroutine around ListenAndServe

tiithansen commented on server/worker/worker.go:
```diff
@@ -0,0 +1,244 @@
+
+func (api *Client) updateDashboards(dashboards []Dashboard) {
+	for _, dashboard := range dashboards {
+		if dashboard.Title == "test dashboard - empty" {
```
I guess this needs to be removed, otherwise it will only update test dashboard. You could also extract this into a dashboard filtering functionality which could be set via cli arg as comma separated list for example.

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,338 @@
+}
+
+func (api *Client) prepare(method string, endpoint string, payload []byte) (*http.Request, error) {
+	bearer := "Bearer " + api.Token
```
I'd inline it or improve variable name

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,338 @@
+
+	req, err := http.NewRequest(method, api.GrafanaUrl+endpoint, bytes.NewBuffer(payload))
+	if err != nil {
+		return nil, errors.New("Failed to create new HTTP Request")
```
add more context to error message. Endpoint and maybe method could be useful to understand from the logs

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,338 @@
+	return false
+}
+
+func (api *Client) getResponseBody(res *http.Response) []byte {
```
is this func unused?

lennarkivimae replied:
It was indeed. Has been removed

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,338 @@
+	}
+}
+
+func (api *Client) AddAnalyticsToDashboards() {
```
I'd reorder functions according to stepdown rule to make code easier to understand

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,338 @@
+	for _, dashboard := range dashboards {
+		if dashboard.Title == "test dashboard - empty" {
+			dashboard.Data["overwrite"] = true
+			dashboard.Data["message"] = "PRO - Auto-add analytics panel"
```
this message should be probably configurable, at least we shouldn't include "PRO" in OS code

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,338 @@
+			if responseAsStruct.Status == "success" {
+				level.Info(api.Logger).Log(
+					"status", "success",
+					"message", "Added analytics to "+dashboard.Title,
```
function name prefix is missing here, but not sure what's the convention in this project

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,338 @@
+
+				break
+			}
+		}
```
I'd extract it into separate function, to make it easier to read on high level

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,338 @@
+			if isAnalyticsPanel(panelMap) {
+				hasAnalyticsPanel = true
+
+				break
```
isn't this a bug related to detecting `largestPanelId` if after adding analytics panel there is some new panel added?

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,338 @@
+	}
+}
+
+func (api *Client) getAnalyticsPanelData(panelId int) map[string]interface{} {
```
`get` prefix is misleading - it creates panel configuration, instead of getting/loading it via the api, as you would incorrectly expect when looking code where it is called from

atsu85 replied:
don't make it instance method of API client - that's just causing even more confusion related to the comment above

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+func (api *Client) updateDashboards(dashboards []Dashboard) {
+	for _, dashboard := range dashboards {
```
I'd extract content of the loop into separate method. 
Probably after that there would be no need for separate method for updating multiple dashboards, as currently there is no other logic besides just for loop

tiithansen commented on server/worker/worker.go:
```diff
@@ -0,0 +1,244 @@
+		for _, panel := range panels {
+			panelMap, ok := panel.(map[string]interface{})
+			if !ok {
+				continue
```
Log error that it was not possible to type cast

tiithansen commented on server/worker/worker.go:
```diff
@@ -0,0 +1,244 @@
+		return panelType == "macropower-analytics-panel"
+	}
+
+	return false
```
```suggestion
	return ok && panelType == "macropower-analytics-panel"
```

Commit:
auto-analytics-adder - increase readability, optimizations

Restructure input order
Add extra params to increase testability / flexibility

tiithansen commented on server/main.go:
```diff
@@ -73,6 +78,28 @@ func main() {
+		Filter:       cli.DashboardFilter,
+	}
+
+	mux.HandleFunc("/work", func(w http.ResponseWriter, r *http.Request) {
```
/work might not be best, maybe something like /patch-dashboards?

Commit:
auto-analytics-adder - remove invalid comment

Commit:
auto-analytics-adder - fix issues with kong usage

Add default value for timeout, for easier usage

Commit:
auto-analytics-adder - cr changes

simplify and increase readability

Commit:
auto-analytics-adder - fix setting content type

Commit:
auto-analytics-adder - Change handler endpoint

Commit:
auto-analytics-adder - revert invalid change

atsu85 commented on server/worker/api.go:
```diff
@@ -0,0 +1,95 @@
+	client := &http.Client{}
+	res, err := client.Do(req)
+	if res.StatusCode != http.StatusOK {
+		return nil, errors.New("Request failed with status: " + strconv.Itoa(res.StatusCode))
```
add more context to error message. Endpoint and maybe method could be useful to understand from the logs

atsu85 commented on server/worker/worker.go:
```diff
@@ -0,0 +1,265 @@
+}
+
+func (api *Client) updateDashboard(dashboard Dashboard, hasFilter bool) {
+	if hasFilter && dashboard.Title != api.Filter {
```
i'd move this if statement around this function call, so the function signature would make more sense (`hasFilter` currently isn't an input to updating the dashboard, but decides if dashboard is updated at all)

Commit:
auto-analytics-adder - cr changes

Move filter check outside of updateDashboards func. Making function
signature correct, update might not of happened at all

increase request failed error information amount for better
understanding why the request failed

atsu85 commented on server/worker/api.go:
```diff
@@ -29,7 +29,9 @@ func (api *Client) Get(endpoint string) ([]byte, error) {
 	res, err := client.Do(req)
 	if res.StatusCode != http.StatusOK {
-		return nil, errors.New("Request failed with status: " + strconv.Itoa(res.StatusCode))
+		errorMessage := fmt.Sprintf("Request failed\nEndpoint: %s\nStatus: %s", endpoint, strconv.Itoa(res.StatusCode))
```
FYI: I'd prefer keeping log messages mostly on single line. I'd consider new lines (with indentation to make following lines of the log record visually related to first log line) in cases when providing additional context (i.e. background information about the error or how to fix the error)
